analysis.ml

```diff
@@ -0,0 +1,58 @@
+open Instr
+
+let successors prog_at program pc =
+  let pc' = pc + 1 in
+  let next = if pc' = Array.length program then [] else [pc'] in
+  let resolve = Instr.generic_resolve prog_at program in
+  match prog_at program pc with
+  | Decl_const _
+  | Decl_mut _
+  | Assign _
+  | Label _
+  | Comment _
+  | Print _ -> next
+  | Goto l | Invalidate (_, l, _) -> [resolve l]
+  | Branch (_e, l1, l2) -> [resolve l1; resolve l2]
+  | Stop -> []
+
+module VarSet = Set.Make(Variable)
+
+(* Perform forward analysis on some code
+ *
+ * init_state : Initial input state to instruction 0
+ * merge      : current state -> input state -> merge state if changed
+ * update     : abstract instruction -> input state -> output state
```

Comment:
We could have a record type `type ('a, 'p) domain = ...` for the `init_state`, `merge` and `update` instructions as they make sense to gether -- and that makes sure that producer and consumer uses consistent naming for these functions, that is the record field names.